Build failure: Stellarium

### Nixpkgs version

- Unstable (25.05)

### Steps to reproduce

- `nix profile install nixpkgs#stellarium`
- Fail

### Can Hydra reproduce this build failure?

Yes, Hydra can reproduce this build failure.

### Link to Hydra build job

https://hydra.nixos.org/build/295424653

### Relevant log output

```console
> /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qstring.h:330:50: error: no type named 'type' in 'struct std::enable_if<false, bool>'
       >   330 |     template <typename T, if_floating_point<T> = true>
       >       |                                                  ^~~~
       > /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qstring.h:336:27: note: candidate: 'template<class T, typename std::enable_if<std::conjunction<std::negation<QtPrivate::treat_as_integral_arg<typename std::remove_cv< <template-parameter-1-1> >::type> >, std::is_convertible<T, QAnyStringView> >::value, bool>::type <anonymous> > QString QString::arg(const T&, int, QChar) const'
       >   336 |     [[nodiscard]] QString arg(const T &a, int fieldWidth = 0, QChar fillChar = u' ') const
       >       |                           ^~~
       > /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qstring.h:336:27: note:   candidate expects 3 arguments, 5 provided
       > /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qstring.h:356:5: note: candidate: 'template<class ... Args> typename std::enable_if<((sizeof... (Args) >= 2) && conjunction_v<std::conjunction<std::negation<QtPrivate::treat_as_integral_arg<typename std::remove_cv<Args>::type> >, std::is_convertible<Args, QAnyStringView> >...>), QString>::type QString::arg(Args&& ...) const'
       >   356 |     arg(Args &&...args) const
       >       |     ^~~
       > /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qstring.h:356:5: note:   template argument deduction/substitution failed:
       > /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qstring.h: In substitution of 'template<class ... Args> typename std::enable_if<((sizeof... (Args) >= 2) && conjunction_v<std::conjunction<std::negation<QtPrivate::treat_as_integral_arg<typename std::remove_cv<Args>::type> >, std::is_convertible<Args, QAnyStringView> >...>), QString>::type QString::arg(Args&& ...) const [with Args = {int, int, char, const int&, char}]':
       > /tmp/nix-build-stellarium-25.1.drv-0/source/src/gui/AngleSpinBox.cpp:530:56:   required from here
       >   530 |                         refText = QString("%1%2°").arg(signPlaceholder).arg(360, 0, 'f', decimalPlaces, ' ');
       >       |                                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
       > /nix/store/cp30fin83aimfljxb5988bkn2micgi04-qtbase-6.9.0/include/QtCore/qstring.h:356:5: error: no type named 'type' in 'struct std::enable_if<false, QString>'
       >   356 |     arg(Args &&...args) const
       >       |     ^~~
       > [ 97%] Building CXX object src/CMakeFiles/stelMain.dir/gui/DateTimeDialog.cpp.o
       > [ 97%] Building CXX object src/CMakeFiles/stelMain.dir/gui/ViewDialog.cpp.o
       > [ 97%] Building CXX object src/CMakeFiles/stelMain.dir/gui/LightPollutionWidget.cpp.o
       > make[2]: *** [src/CMakeFiles/stelMain.dir/build.make:2558: src/CMakeFiles/stelMain.dir/gui/AngleSpinBox.cpp.o] Error 1
       > make[2]: *** Waiting for unfinished jobs....
       > make[1]: *** [CMakeFiles/Makefile2:3608: src/CMakeFiles/stelMain.dir/all] Error 2
       > make: *** [Makefile:156: all] Error 2
```

### Additional context

A Qt version issue maybe?
It seems to build with an older version of it at any rate [here](https://hydra.nixos.org/build/295362650/log)

### System metadata

 - system: `"x86_64-linux"`
 - host os: `Linux 6.14.3, NixOS, 25.05 (Warbler), 25.05.20250421.c11863f`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.28.1`
 - channels(root): `""`
 - nixpkgs: `/nix/store/kcmmd6alr3lx56vkf72503h3pxgf6iv4-source`

### Notify maintainers

@kilianar

---

**Note for maintainers:** Please tag this issue in your pull request description. (i.e. `Resolves #ISSUE`.)


### I assert that this issue is relevant for Nixpkgs

- [x] I assert that this is a bug and not a support request.
- [x] I assert that this is not a [duplicate of an existing issue](https://github.com/NixOS/nixpkgs/issues?q=is%3Aopen+is%3Aissue+label%3A%220.kind%3A+build+failure%22). 
- [x] I assert that I have read the [NixOS Code of Conduct](https://github.com/NixOS/.github/blob/master/CODE_OF_CONDUCT.md) and agree to abide by it.

### Is this issue important to you?

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/applications/science/astronomy/stellarium/default.nix
+++ b/pkgs/applications/science/astronomy/stellarium/default.nix
@@ -24,6 +24,7 @@
   xvfb-run,
   gitUpdater,
   md4c,
+  fetchpatch,
 }:
 
 stdenv.mkDerivation (finalAttrs: {
@@ -37,19 +38,27 @@ stdenv.mkDerivation (finalAttrs: {
     hash = "sha256-rbnGSdzPuFdSqWPaKtF3n4oLZ9l+4jX7KtnmcrTvwbs=";
   };
 
-  postPatch =
-    ''
-      substituteInPlace CMakeLists.txt \
-        --replace-fail 'CPMAddPackage(NAME md4c' \
-                  'CPMFindPackage(NAME md4c'
-    ''
-    + lib.optionalString stdenv.hostPlatform.isDarwin ''
-      substituteInPlace CMakeLists.txt \
-        --replace-fail 'SET(CMAKE_INSTALL_PREFIX "''${PROJECT_BINARY_DIR}/Stellarium.app/Contents")' \
-                  'SET(CMAKE_INSTALL_PREFIX "${placeholder "out"}/Applications/Stellarium.app/Contents")'
-      substituteInPlace src/CMakeLists.txt \
-        --replace-fail "\''${_qt_bin_dir}/../" "${qtmultimedia}/lib/qt-6/"
-    '';
+  patches = [
+    # Patch from upstream to fix compilation with Qt 6.9
+    (fetchpatch {
+      url = "https://github.com/Stellarium/stellarium/commit/bbcd60ae52b6f1395ef2390a2d2ba9d0f98db548.patch";
+      hash = "sha256-9VaqLASxn1udUApDZRI5SCqCXNGOHUcdbM+pKhW8ZAg=";
+    })
+
+    # Upstream patch to support building with a locally provided md4c package
+    (fetchpatch {
+      url = "https://github.com/Stellarium/stellarium/commit/972c6ba72f575964fbf2049a22d51b4d1fd3983c.patch";
+      hash = "sha256-ef1Jw5NeT0KLVKQt7VcvQh83n2ujMFK+Nv0165ZQ2r8=";
+    })
+  ];
+
+  postPatch = lib.optionalString stdenv.hostPlatform.isDarwin ''
+    substituteInPlace CMakeLists.txt \
+      --replace-fail 'SET(CMAKE_INSTALL_PREFIX "''${PROJECT_BINARY_DIR}/Stellarium.app/Contents")' \
+                'SET(CMAKE_INSTALL_PREFIX "${placeholder "out"}/Applications/Stellarium.app/Contents")'
+    substituteInPlace src/CMakeLists.txt \
+      --replace-fail "\''${_qt_bin_dir}/../" "${qtmultimedia}/lib/qt-6/"
+  '';
 
   nativeBuildInputs = [
     cmake

